core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -195,6 +195,201 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
+
+    // would advance to NODE_LOCAL locality if timer wasn't reset above
+    // Verifying this node local task is not accepted
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.isEmpty)
+  }
+
+  test("SPARK-18886 - delay scheduling timer is reset when it accepts all resources offered when" +
+    "isAllFreeResources = true") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    // timer is reset when tsm accepts all free resources offered to it
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = true)
+      .flatten.length === 1)
+
+    // would advance to NODE_LOCAL locality if timer wasn't reset above
+    // Verifying this node local task is not accepted
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.isEmpty)
+  }
+
+  test("SPARK-18886 - partial resource offers (isAllFreeResources = false) reset " +
+    "time if last full resource offer (isAllResources = true) was accepted as well as any " +
+    "following partial resource offers") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = true)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec1", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.length === 1)
+
+    clock.advance(advanceAmount)
+    assert(taskScheduler
+      .resourceOffers(
+        IndexedSeq(WorkerOffer("exec2", "host1", 1)),
+        isAllFreeResources = false)
+      .flatten.isEmpty)
+  }
+
+
+  // This tests two cases
+  // 1. partial resource offer doesn't reset timer after full resource offer had rejected resources
+  // 2. partial resource offer doesn't reset timer after partial resource offer
+  //    had rejected resources
+  test("SPARK-18886 - partial resource offers (isAllFreeResources = false) do not reset " +
+    "time if any offer was rejected since last full offer was fully accepted") {
+    val clock = new ManualClock()
+    val taskScheduler = setupTaskScheduler(clock)
+    val advanceAmount = 2000
+
+    // case 1 from test description above
+    assert(taskScheduler
+      .resourceOffers(
```

Comment:
I'd like to see more comments to explain what happens when we call `resourceOffers`

For example, this one: `It's not a node-local resource, no task is scheduled due to delay scheduling`.